File: Diagonal array.java
// Time complexity 0(m*n)
// space complexity 0(1)
// Code successfully executed in leetcode
class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
       
        if(matrix == null || matrix.length == 0) return new int []{};
        int m = matrix.length;
        int n = matrix[0].length;
        int result[] = new int [m*n];
        
        int i=0;
        int r=0;
        int c=0;
        int dir =1;
        
        while(i < m*n){
            result[i] = matrix[r][c];
            if(dir == 1){
                if(c == n-1){
                    r++;
                    dir = -1;
                }else if (r==0){
                    c++;
                    dir = -1;
                    
                }else {
                    r--;
                    c++;   
                }
            } else {
                if(r == m-1){
                    c++;
                    dir = 1;
                }else if (c==0){
                    r++;
                    dir = 1;
                    
                }else {
                    r++;
                    c--;   
                }
                
            }
            i++;
        }
        return result;
         
    }
}

File: diagonalArray.cpp
// Time Complexity :O(m*n) where m*n in the number elements in the matrix
// Space Complexity : O(1)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        int dir  = 1;
        int r = 0,c =0;
        vector<int> ans;
        int m = mat.size();
        int n = mat[0].size();
        int i = 0;
        while(i< m*n){
            ans.push_back(mat[r][c]);
            if(dir == 1){
                if(c == n-1){ //2,2
                dir = -1;
                r++;
            }
            else if(r == 0){
                dir = -1;
                c++;
            }
            else{
                c++;
                r--;
            }
            }
            else{
                if(r == m-1){
                    dir = 1;
                    c++;
                }
                else if(c == 0){
                    dir = 1;
                    r++;
                }
                else{
                    c--;
                    r++;
                }
            }
            i++;
        }
        return ans;
    }
};